pkg/clusteragent/admission/controllers/webhook/controller_base.go

```diff
@@ -186,9 +186,11 @@ func (c *controllerBase) EnabledWebhooks() []Webhook {
 func (c *controllerBase) enqueueOnLeaderNotif(stop <-chan struct{}) {
 	for {
 		select {
-		case <-c.isLeaderNotif:
-			log.Infof("Got a leader notification, enqueuing a reconciliation for %q", c.config.getWebhookName())
-			c.triggerReconciliation()
+		case <-c.leadershipStateNotif:
+			if c.isLeaderFunc() {
```

Comment:
General question: Are we certain that the `isLeaderFunc()` return value is always updated before receiving the `leadershipStateNotif` notification?
What I’d like to confirm is that there’s no potential race condition between these two lines. For example, the notification might be sent because the DCA instance became the leader, but this information hasn’t yet propagated to the `c.isLeaderFunc()` output. This could result in skipping the execution of `c.triggerReconciliation()`.

Would it make sense to add a log in the else case, stating: “Received a notification but not recognized as leader yet”?

Reply:
The notification is never sent before the leadership status is updated fully and propagated to `isLeader`

[Here](https://github.com/DataDog/datadog-agent/blob/290d4bac7776f461dce431287925ab86fcd74aa0/pkg/util/kubernetes/apiserver/leaderelection/leaderelection_engine.go#L155-L157) is where `notify` is called. You can see that we call `updateLeaderIdentity` before we notify subscribers.

Finally, `IsLeader` has [a very simple implementation](https://github.com/DataDog/datadog-agent/blob/290d4bac7776f461dce431287925ab86fcd74aa0/pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go#L279-L282). Once we `updateLeaderIdentity`, the change is instantly visible in `isLeader` func result.

Reply:
> Would it make sense to add a log in the else case, stating: “Received a notification but not recognized as leader yet”?

If it is not clear, the notification is not to tell the subscriber that the instance has become a leader. But rather to tell it that a change has occurred on the leadership state. 

The subscriber is responsible to handle the logic based on the usecase, and `isLeaderFunc` is provided to determine the live state.

Reply:
To add, I already added unit tests for the subscription feature which validates that the change will have already been propagated to `isLeaderFunc` when the notification is sent. 

https://github.com/DataDog/datadog-agent/blob/290d4bac7776f461dce431287925ab86fcd74aa0/pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go#L237-L239